### PR TITLE
Added the ability to fine-tune on data

### PR DESCRIPTION
**Pull request by ivorsimpson**
This required a bit of tweaking in terms of the synthetic noise distribution and playing around with the model. Seems to work quite nicely now

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -46,21 +47,50 @@ def loss_fn(y_true, y_pred):
     return tf.reduce_mean(nll)
 
 
-def crop_3d():
-    crop_size = 20
+class ReparamTrickLayer(keras.layers.Layer):
```
Comment (ashmcmn) [inline train.py:162]: What is this function doing?

Comment (ashmcmn) [reply train.py:162]: Generates a sample of oef and dbv values by using the predictions of mean and std dev for the signal generation layer?

File: train.py [hunk 2]
```diff
@@ -46,21 +47,50 @@ def loss_fn(y_true, y_pred):
...
+    def call(self, input, *args, **kwargs):
+        oef_sample = input[:, :, :, :, 0] + tf.random.normal(tf.shape(input[:, :, :, :, 0])) * tf.exp(
+            input[:, :, :, :, 1])
+        dbv_sample = input[:, :, :, :, 2] + tf.random.normal(tf.shape(input[:, :, :, :, 0])) * tf.exp(
```
Comment (ashmcmn) [inline train.py:167]: Should this be
 `dbv_sample = input[:, :, :, :, 2] + tf.random.normal(tf.shape(input[:, :, :, :, 2]))` * ...

Comment (ashmcmn) [reply train.py:167]: Found it in [the paper](https://arxiv.org/pdf/1312.6114.pdf): 2.4 The reparameterization trick

File: train.py (hunk 1 above)
Comment (ashmcmn) [reply train.py:162]: Found it in [the paper](https://arxiv.org/pdf/1312.6114.pdf): 2.4 The reparameterization trick

File: signals.py [hunk 3]
```diff
@@ -91,14 +100,25 @@ def call(self, input, *args, **kwargs):
             else:
                 noise_weights = 1
 
-            stdd = abs(tf.math.reduce_mean(signal, axis=1)) / self._snr  # calculate standard deviation for each signal
-            stdd = tf.repeat(tf.expand_dims(stdd, 1), 11, axis=1)
+            #stdd = abs(tf.math.reduce_mean(signal, axis=1)) / self._snr  # calculate standard deviation for each signal
+            #stdd = tf.repeat(tf.expand_dims(stdd, 1), 11, axis=1)
+
+            # Use the std-dev as observed from the real normalised data
+            stdd = tf.random.uniform((signal.shape.as_list()[0],1), 0.07, 0.09)
```
Comment (ashmcmn) [inline signals.py]: Could this be automated/loaded from preprocessing?

File: train.py (hunk 2 above)
Comment (ivorsimpson) [reply train.py:167]: I'm just taking the shape of it and it's the same for all of them. It'd be cleaner if we just saved the shape for the samples and passed it into both of them

File: signals.py (hunk 3 above)
Comment (ivorsimpson) [reply signals.py]: This is an old version - the new noise generation code is different

File: data_preprocessing.py [hunk 4]
```diff
@@ -2,33 +2,66 @@
...
+    corner2 = data[-corner_vals:-1, -corner_vals:-1, :, :]
+    corner3 = data[-corner_vals:-1, 1:corner_vals, :, :]
+    corner4 = data[1:corner_vals, -corner_vals:-1, :, :]
+    corner = np.concatenate([corner, corner2, corner3, corner4], 0)
     # Reshape
     corner = corner.reshape(-1, 11)
+
+    corner_std = np.std(corner[:, :], 0)
+    snr = within_mask_mean / corner_std
+    norm_snr = snr / snr[3]
```
Comment (ashmcmn) [inline data_preprocessing.py]: Maybe we could then save this for use where aforementioned stdd range is hardcoded

File: train.py [hunk 5]
```diff
@@ -136,36 +281,52 @@ def save_predictions(model, data, filename):
         x = np.reshape(x, (-1, 1, 1, 1, 11))
         y = np.reshape(y, (-1, 1, 1, 1, 2))
 
-    optimiser = tf.keras.optimizers.Adam()
+    train_x = x[:-1000, ...]
```
Comment (ashmcmn) [inline train.py]: Should this be done on a proportion instead of hard constant

File: train.py [hunk 6]
```diff
@@ -136,36 +281,52 @@ def save_predictions(model, data, filename):
...
+    valid_y = y[-1000:, ...]
+
+    optimiser = tf.keras.optimizers.Adam(learning_rate=1e-3)
+
+    # system_constants = get_constants(params)
     model = create_model(train_conv)
-    model.compile(optimiser, loss=loss_fn)
+    model.compile(optimiser, loss=[loss_fn, None], metrics=[[oef_metric, dbv_metric], None])
 
     es = tf.keras.callbacks.EarlyStopping(monitor='val_loss', patience=10, verbose=1)
```
Comment (ashmcmn) [inline train.py:381]: This could be removed if no longer used

File: train.py [hunk 7]
```diff
@@ -9,38 +9,62 @@
...
-            return keras.layers.Conv3D(no_units, kernel_size=(1, 1, 1), activation=activation)
-        else:
-            return keras.layers.Dense(no_units, activation=activation)
-
     net = create_layer(no_units)(net)
+    net = add_normalizer(net)
     if system_constants is not None and i == 0:
         net = keras.layers.Multiply()([net, keras.layers.Reshape((1, 1, 1, -1))(const_net)])
 
     for i in range(1):
```
Comment (ashmcmn) [inline train.py:59]: Why is this looped but only once?

**Review by ashmcmn — APPROVED** [review]

**Pull request merged**